samcli/commands/init/init_templates.py

```diff
@@ -118,23 +120,72 @@ def _shared_dir_check(self, shared_dir):
         shared_dir = global_cfg.config_dir
         if not self._shared_dir_check(shared_dir):
+            # Nothing we can do if we can't access the shared config directory, use bundled.
             return
         expected_path = os.path.normpath(os.path.join(shared_dir, self._repo_name))
-        if self._should_clone_repo(expected_path):
+        if self._template_directory_exists(expected_path):
+            self._overwrite_existing_templates(shared_dir, expected_path)
+        else:
+            # simply create the app templates repo
+            self._clone_new_app_templates(shared_dir, expected_path)
+        self.clone_attempted = True
+
+    def _overwrite_existing_templates(self, shared_dir, expected_path):
+        self.repo_path = expected_path
+        # workflow to clone a copy to a new directory and overwrite
+        with osutils.mkdir_temp(ignore_errors=True) as tempdir:
             try:
+                expected_temp_path = os.path.normpath(os.path.join(tempdir, self._repo_name))
+                LOG.info("\nCloning app templates from %s", self._repo_url)
                 subprocess.check_output(
-                    [self._git_executable(), "clone", self._repo_url], cwd=shared_dir, stderr=subprocess.STDOUT
+                    [self._git_executable(), "clone", self._repo_url, self._repo_name],
+                    cwd=tempdir,
+                    stderr=subprocess.STDOUT,
                 )
+                # Now we need to delete the old repo and move this one.
+                try:
+                    LOG.debug("Removing old templates from %s", str(expected_path))
+                    shutil.rmtree(expected_path)
+                    LOG.debug("Copying templates from %s to %s", str(expected_temp_path), str(expected_path))
+                    shutil.copytree(expected_temp_path, expected_path, ignore=shutil.ignore_patterns("*.git"))
+                except (OSError, shutil.Error) as ex:
+                    # UNSTABLE STATE - it's difficult to see how this scenario could happen except weird permissions, user will need to debug
+                    raise AppTemplateUpdateException(
+                        "Unstable state when updating app templates. Check that you have permissions to create/delete files in the AWS SAM shared directory or file an issue at https://github.com/awslabs/aws-sam-cli/issues"
+                    )
+                self.repo_path = expected_path
+            except OSError as ex:
+                LOG.warning("WARN: Could not clone app template repo.", exc_info=ex)
+            except subprocess.CalledProcessError as clone_error:
+                output = clone_error.output.decode("utf-8")
+                if "not found" in output.lower():
+                    click.echo("WARN: Could not clone app template repo.")
+
+    def _clone_new_app_templates(self, shared_dir, expected_path):
+        with osutils.mkdir_temp(ignore_errors=True) as tempdir:
```

Comment:
if we are using ignore_errors, i also suggest we add a callback for `onerror` to log a debug message.

Reply:
Done in latest change